core/src/main/java/com/scalar/db/transaction/consensuscommit/Coordinator.java

```diff
@@ -125,30 +125,23 @@ public void putState(Coordinator.State state) throws CoordinatorException {
     put(put);
   }
 
-  // TODO: This method should be separated into `putStateForGroupCommitWithParentId()` and
-  //       `putStateForGroupCommitWithFullId()` so whether the id is a parent ID or a full ID is
-  //       clear.
   void putStateForGroupCommit(
```

Comment:
Sorry, this is not directly related to this PR, but I'm wondering if we could separate this group commit related logic to another class. For example, can we create a subclass of this `Coordinator` class, something like `CoordinatorWithGroupCommit`, and move the group commit related logic to this class?

Reply:
Agree. It is worth thinking about it.

Reply:
@brfrn169 @feeblefakie Thanks for the feedback!

I first thought the same thing like the relationship of `CommitHandler` and `CommitHandlerWithGroupCommit`, either of them is used based on the configuration. But, in case of `Coordinator`, it could need to handle full IDs generated for the group commit even if the group commit feature is disabled as follows:

1. An application starts with the group commit feature is enabled
2. A transaction creates a PREPARE state record that has a full transaction ID and gets crashed w/o commit or abort
3. The application restarts with the group commit feature is disabled
4. A transaction reads the PREPARE state record and tries to check if it's committed or aborted via `Coordinator` 
5. But the `Coordinator` can handle only normal transaction IDs. So, `Coordinator` needs to be able to handle the full ID despite the group commit feature is disabled

Of course, we can create `CoordinatorWithGroupCommit` that inherits `Coordinator` and is able to handle both normal transaction ID and group commit related transaction ID, and always use it:
- `CoordinatorWithGroupCommit.getState(String id)` will be as follows
  - If `id` is for normal transaction ID, it's delegated to `super.getState(id)`
  - Otherwise, the `id` is handled with `CoordinatorWithGroupCommit.getStateForGroupCommit(id)`
- Transaction Managers always use `CoordinatorWithGroupCommit` instead of `Coordinator`. In other word, `Coordinator` will be like abstract class.

I wasn't sure this was better. But if you think it's better than the current fat `Coordinator`, I'll handle the refactoring as another PR. What do you think?

Reply:
Okay, understood. I didn't assume the situation for disabling after enabling and the group commit feature. Let's leave it as it is.

BTW, do we really need to assume the situation? I feel that makes the things complicated (the code, tests, etc.).

One more question, if we have mixed multiple transaction manager instances that enable and disable the group commit feature, does that work?

Reply:
> BTW, do we really need to assume the situation? I feel that makes things complicated (the code, tests, etc.).

Good point. I think it's a very corner case, but it might potentially occur. If the transaction state is committed before crashing in the above example case, the committed transaction might be rolled back by `Coordinator` which can't find the transaction with full ID after the restart. It seems a critical case and should be prevented as much as possible.

Also, I agree with your concern. In this case, we need to only consider the scenario where `Coordinator.getState(fullID)` is called with the group committer disabled, so probably don't need so much complicated things. (I'll add the test case later...)

> if we have mixed multiple transaction manager instances that enable and disable the group commit feature, does that work?

Yes. I think it works as far as a transaction started with a transaction manager that enables the group commit feature isn't handled by another transaction manager that disables the feature, or vise versa.
- Transaction Manager (TM1) with the group commit feature enabled starts Transaction 1
- There is another Transaction Manager (TM2) with the group commit feature disabled 
- Transaction 1 needs to be handled by TM1 always, not by TM2